Accurately annotate Error system with [[noreturn]]

alexreinking opened the pull request:
This PR splits the ErrorReport type into two types:

1. `WarningReport`, which does not throw/abort on destruction.
2. `ErrorReport`, which always throws/aborts on destruction, and is annotated with `[[noreturn]]`.

This makes `internal_error` a valid statement to use at the end of a function that would otherwise fail to return.

Also applies `HALIDE_ALWAYS_INLINE` to the `operator<<` overload because it was getting specialized for every size of `const char *` literal.

abadams commented:
Should CompileTimeErrorReporter::error also be annotated?

abadams commented:
Yes, because it needs to be crystal clear that the method must not return, or there is UB. I suspect the only downstream user is our own python bindings.

alexreinking commented:
> Yes, because it needs to be crystal clear that the method must not return, or there is UB. I suspect the only downstream user is our own python bindings.

I added the annotation and rebased on main. Hope to get this merged soon!
